libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+		eprintf ("XMM%i %016"PFMT64x" %016"PFMT64x"\n", x, xmm[x].High, xmm[x].Low);
+	}
+	// show Ymm regs
+	nymm = GetAVX (hThread, xmm, ymm);
+	if (nymm) {
+		for (x = 0; x < nymm; x++) {
+			eprintf ("Ymm%d: %016"PFMT64x" %016"PFMT64x" %016"PFMT64x" %016"PFMT64x"\n", x, ymm[x].High, ymm[x].Low, xmm[x].High, xmm[x].Low );
+		}
+	}
+}
+
+int w32_reg_read(RDebug *dbg, int type, ut8 *buf, int size) {
+	bool showfpu = false;
+	if (type < -1) {
+		showfpu = true; // hack for debugging
+		type = -type;
+	}
+	DWORD flags = THREAD_SUSPEND_RESUME | THREAD_GET_CONTEXT;
+	if (dbg->bits == R_SYS_BITS_64) {
+		flags |= THREAD_QUERY_INFORMATION;
+	}
+	HANDLE th = OpenThread (flags, FALSE, dbg->tid);
+	if (th == (HANDLE)NULL) {
+		r_sys_perror ("w32_reg_read/OpenThread");
+		return 0;
+	}
+	// Always suspend
+	if (suspend_thread (th, dbg->bits) == -1) {
+		CloseHandle (th);
+		return 0;
+	}
+	size = get_thread_context (th, buf, size, dbg->bits);
+	if (showfpu) {
+		printwincontext (th, (CONTEXT *)buf);
+	}
+	// Always resume
+	if (resume_thread (th, dbg->bits) == -1) {
+		size = 0;
+	}
+	CloseHandle (th);
+	return size;
+}
+
+int w32_reg_write(RDebug *dbg, int type, const ut8 *buf, int size) {
+	DWORD flags = THREAD_SUSPEND_RESUME | THREAD_SET_CONTEXT;
+	if (dbg->bits == R_SYS_BITS_64) {
+		flags |= THREAD_QUERY_INFORMATION;
+	}
+	HANDLE th = OpenThread (flags, FALSE, dbg->tid);
+	if (th == (HANDLE)NULL) {
+		r_sys_perror ("w32_reg_write/OpenThread");
+		return false;
+	}
+	// Always suspend
+	if (suspend_thread (th, dbg->bits) == -1) {
+		CloseHandle (th);
+		return false;
+	}
+	bool ret = set_thread_context (th, buf, size, dbg->bits);
+	// Always resume
+	if (resume_thread (th, dbg->bits) == -1) {
+		ret = false;
+	}
+	CloseHandle (th);
+	return ret;
+}
+
+int w32_attach(RDebug *dbg, int pid) {
+	RIOW32Dbg *rio = dbg->user;
+	HANDLE ph = OpenProcess (PROCESS_ALL_ACCESS, FALSE, pid);
+	if (ph == (HANDLE)NULL) {
+		return -1;
+	}
+	if (!DebugActiveProcess (pid)) {
+		CloseHandle (ph);
+		return -1;
+	}
+	RList *threads = r_list_new ();
+	if (!threads) {
+		CloseHandle (ph);
+		return -1;
+	}
+	threads = w32_thread_list (dbg, pid, threads);
+	if (threads->length == 0) {
+		r_list_free (threads);
+		CloseHandle (ph);
+		return -1;
+	}
+	int tid = ((RDebugPid *)threads->head->data)->pid;
+	r_list_free (threads);
+	rio->ph = ph;
+	rio->debug = true;
+	return tid;
+}
+
+int w32_detach(RDebug *dbg, int pid) {
+	if (pid == -1) {
+		return false;
+	}
+	if (dbg->pid == pid) {
+		RIOW32Dbg *rio = dbg->user;
+		bool ret;
+		if (rio->debug) {
+			ret = DebugActiveProcessStop (pid);
+		}
+		CloseHandle (rio->ph);
+		rio->ph = NULL;
+		rio->debug = false;
+		return ret;
+	}
+	return false;
+}
+
+int w32_select(RDebug *dbg, int pid, int tid) {
+	RIOW32Dbg *rio = dbg->user;
+	if (rio->ph != (HANDLE)NULL) {
+		return true;
+	}
+	// hack to support w32dbg:// and attach://
+	DEBUG_EVENT de;
+	de.dwProcessId = pid;
+	de.dwThreadId = tid;
+	bool cont = true;
```

Comment:
there are 3 exit points of this loop, 2 with breaks, one with conditional variable, i think it will be good to just have only one way to quit the loop

Reply:
Yes I agree that loop is pretty ugly.